File: src/SuperClock.cpp
/*
 * SuperClock.cpp — shared NTP-domain getters + beat math.
 *
 * Methods that touch Link or that vary by platform (mutators, Link-aware
 * getters, Link-clock RPC, audio-thread Link Audio paths) live in the
 * platform .cpp files. This file holds only the pieces that are
 * identical on native and WASM: pure reads of the SuperClockState
 * atomics + the NTP-domain beat math composed from them.
 */
#include "SuperClock.h"
#include "clock_math.h"
#include "shared_memory.h"
#include "synth/common/shm_scope_stream.hpp"  // g_engine_frames (stream anchor)

#include <atomic>
#include <cmath>
#include <cstdio>
#include <cstring>

using supersonic::bitsToDouble;

std::atomic<SuperClock*> g_active_superclock{nullptr};

// ── Sample clock (sample position ↔ wall-clock DAC time) ───────────────

void SuperClock::bindSampleClockToShm(uint8_t* region) {
    mSampleClockRegion = region;
}

void SuperClock::advanceEngineFrames(double samplePosition) {
    g_engine_frames.store(static_cast<uint64_t>(samplePosition),
                          std::memory_order_relaxed);
}

// Seqlock writer: odd seq, release fence (orders the odd store before the
// field stores for any reader that sees them), relaxed atomic field stores,
// even seq with release. Fields are atomics so no read tears; the seq guards
// cross-field consistency. Audio thread only — single writer by construction.
void SuperClock::publishSampleClock(double samplePosition, double sampleRate,
                                    double renderNtp,
                                    uint32_t outputLatencyFrames) {
    const uint64_t frames = static_cast<uint64_t>(samplePosition);
    // Streams anchor their per-block writes here whether or not a shm
    // region is bound (headless/unit contexts).
    g_engine_frames.store(frames, std::memory_order_relaxed);

    uint8_t* sc = mSampleClockRegion;
    if (!sc || sampleRate <= 0.0) return;
    auto* seq = reinterpret_cast<std::atomic<uint32_t>*>(sc + SAMPLE_CLOCK_SEQ);
    auto* sr  = reinterpret_cast<std::atomic<uint32_t>*>(sc + SAMPLE_CLOCK_SAMPLE_RATE);
    auto* fr  = reinterpret_cast<std::atomic<uint64_t>*>(sc + SAMPLE_CLOCK_ENGINE_FRAMES);
    auto* nb  = reinterpret_cast<std::atomic<uint64_t>*>(sc + SAMPLE_CLOCK_DAC_NTP);
    auto* lat = reinterpret_cast<std::atomic<uint32_t>*>(sc + SAMPLE_CLOCK_OUT_LATENCY);

    // The anchor is speaker-time: render NTP plus the device output latency
    // (Link convention — "host time at speaker").
    const double dacNtp = renderNtp
        + static_cast<double>(outputLatencyFrames) / sampleRate;
    const uint64_t ntpBits = supersonic::doubleToBits(dacNtp);

    const uint32_t s = seq->load(std::memory_order_relaxed);
    seq->store(s + 1, std::memory_order_relaxed);
    std::atomic_thread_fence(std::memory_order_release);
    sr->store(static_cast<uint32_t>(sampleRate), std::memory_order_relaxed);
    fr->store(frames, std::memory_order_relaxed);
    nb->store(ntpBits, std::memory_order_relaxed);
    lat->store(outputLatencyFrames, std::memory_order_relaxed);
    seq->store(s + 2, std::memory_order_release);
}

// ── NTP-domain getters ──────────────────────────────────────────────────

double SuperClock::getBeatOriginNtp() const {
    const SuperClockState* s = state();
    if (!s) return 0.0;
    return bitsToDouble(s->beat_origin_ntp.load(std::memory_order_relaxed));
}

double SuperClock::getIsPlayingAtNtp() const {
    const SuperClockState* s = state();
    if (!s) return 0.0;
    return bitsToDouble(s->is_playing_at_ntp.load(std::memory_order_relaxed));
}

// tempo / transport read from the shared SuperClockState mirror — identical on
// every build. The mirror is written by setBpm/setIsPlaying and, on native, kept
// in sync with Link's converged value by Link's tempo/transport callbacks, so a
// relaxed atomic read here is RT-safe and consistent on every build.
double SuperClock::getBpm() const {
    const SuperClockState* s = state();
    if (!s) return 120.0;
    return bitsToDouble(s->bpm.load(std::memory_order_relaxed));
}

bool SuperClock::isPlaying() const {
    const SuperClockState* s = state();
    if (!s) return false;
    return s->is_playing.load(std::memory_order_relaxed) != 0u;
}

// ── NTP-domain beat math ────────────────────────────────────────────────
// Pure functions of (bpm, beat_origin). Independent atomic reads — no
// multi-field coherence guarantee. App-thread callers don't need it;
// audio-thread callers should use captureSessionState() instead.

double SuperClock::beatAtTime(double ntpSeconds, double quantum) const {
    (void)quantum;
    return supersonic::beatAt(ntpSeconds, getBeatOriginNtp(), getBpm());
}

double SuperClock::phaseAtTime(double ntpSeconds, double quantum) const {
    return supersonic::wrapPhase(beatAtTime(ntpSeconds, quantum), quantum);
}

double SuperClock::timeAtBeat(double beat, double quantum) const {
    (void)quantum;
    return supersonic::timeAtBeat(beat, getBeatOriginNtp(), getBpm());
}

// ── Cross-platform clock metrics ────────────────────────────────────────
// Reads the SuperClockState SAB mirror directly (relaxed atomics) rather than
// getBpm()/isPlaying(), which on native+Link route through the locking
// captureAppSessionState() (not RT-safe). The mirror is the SAB region on WASM
// and a Link-callback-synced private mirror on native, so a direct read is
// RT-safe and consistent on both builds. Fixed-point encoding matches the
// link_* clock slots so the same display formats apply.

void SuperClock::publishClockMetrics(PerformanceMetrics* m, double ntpNow, double quantum) {
    if (!m) return;

    const SuperClockState* s = state();
    if (!s) return;

    // Acquire-load bpm BEFORE origin: applyTempoChange stores the origin first
    // and releases bpm, so a new bpm here guarantees the matching origin.
    const double bpm = bitsToDouble(s->bpm.load(std::memory_order_acquire));
    const double beatOrigin = bitsToDouble(s->beat_origin_ntp.load(std::memory_order_relaxed));
    const bool playing = s->is_playing.load(std::memory_order_relaxed) != 0u;

    const double beat = (bpm > 0.0) ? supersonic::beatAt(ntpNow, beatOrigin, bpm) : 0.0;
    const double phase = supersonic::wrapPhase(beat, quantum);

    m->clock_tempo_mbpm.store(bpm > 0.0 ? static_cast<uint32_t>(bpm * 1000.0 + 0.5) : 0u,
                              std::memory_order_relaxed);
    m->clock_beat_centi.store(beat > 0.0 ? static_cast<uint32_t>(beat * 100.0) : 0u,
                              std::memory_order_relaxed);
    m->clock_phase_centi.store(phase > 0.0 ? static_cast<uint32_t>(phase * 100.0) : 0u,
                               std::memory_order_relaxed);
    m->clock_playing.store(playing ? 1u : 0u, std::memory_order_relaxed);
}

File: js/lib/superclock.js
// SPDX-License-Identifier: MIT OR GPL-3.0-or-later
// Copyright (c) 2025 Sam Aaron

/**
 * SuperClock — single JS-side authority for session state and time.
 * Mirrors the C++ SuperClock API (src/SuperClock.h).
 *
 * Owns:
 *   - Session state: BPM, transport, beat origin. SAB-mode writes the
 *     SuperClockState region directly via BigInt64 atomics; PM-mode
 *     posts to the worklet which writes the same region.
 *   - Time: ntp_start_time / drift / clock-offset via a composed
 *     private NTPTiming helper.
 */

import { NTPTiming } from './ntp_timing.js';
import {
  SC_BPM_I64,
  SC_BEAT_ORIGIN_NTP_I64,
  SC_IS_PLAYING_AT_NTP_I64,
  SC_IS_PLAYING_I32,
  SuperClockMessageType,
  doubleToBits, bitsToDouble,
} from './superclock_protocol.js';

export class SuperClock {
  #mode;
  #workletPort;
  #audioContext;

  // SAB views over the SuperClockState region (same bytes, two strides).
  #sabBigInt;  // BigInt64Array — for the three double fields
  #sabInt32;   // Int32Array    — for the is_playing field

  // PM-mode local copies. In SAB mode these are also kept up-to-date so
  // getters have a fallback before initSharedViews has run.
  #localBpm = 120.0;
  #localBeatOriginNtp = 0.0;
  #localIsPlaying = false;
  #localIsPlayingAtNtp = 0.0;

  #ntp;

  /**
   * @param {Object} options
   * @param {'sab'|'postMessage'} options.mode
   * @param {AudioContext} options.audioContext
   * @param {MessagePort} [options.workletPort] — required in PM mode
   */
  constructor(options = {}) {
    this.#mode = options.mode || 'sab';
    this.#workletPort = options.workletPort || null;
    this.#audioContext = options.audioContext || null;
    this.#ntp = new NTPTiming({
      mode: this.#mode,
      audioContext: options.audioContext,
      workletPort: this.#workletPort,
    });
  }

  /**
   * Initialize SAB views for session state and time.
   * @param {SharedArrayBuffer} sharedBuffer
   * @param {number} ringBufferBase
   * @param {Object} bufferConstants — must include SUPERCLOCK_STATE_START / _SIZE
   */
  initSharedViews(sharedBuffer, ringBufferBase, bufferConstants) {
    if (this.#mode === 'sab') {
      const base = ringBufferBase + bufferConstants.SUPERCLOCK_STATE_START;
      const size = bufferConstants.SUPERCLOCK_STATE_SIZE;
      this.#sabBigInt = new BigInt64Array(sharedBuffer, base, size / 8);
      this.#sabInt32  = new Int32Array(sharedBuffer, base, size / 4);
    }
    this.#ntp.initSharedViews(sharedBuffer, ringBufferBase, bufferConstants);
  }

  /** @param {MessagePort} port */
  setWorkletPort(port) { this.#workletPort = port; }

  updateAudioContext(audioContext) {
    this.#audioContext = audioContext;
    this.#ntp.updateAudioContext(audioContext);
  }

  // ── Time / drift surface (delegates to private NTPTiming) ──────────────

  async initialize() { await this.#ntp.initialize(); }
  resync()           { this.#ntp.resync(); }
  startDriftTimer()  { this.#ntp.startDriftTimer(); }
  stopDriftTimer()   { this.#ntp.stopDriftTimer(); }
  updateDriftOffset(){ this.#ntp.updateDriftOffset(); }
  getDriftOffset()   { return this.#ntp.getDriftOffset(); }
  getNTPStartTime()  { return this.#ntp.getNTPStartTime(); }
  getClockOffset()   { return this.#ntp.getClockOffset(); }
  setClockOffset(s)  { this.#ntp.setClockOffset(s); }
  reset()            { this.#ntp.reset(); }

  // ── Session mutators ───────────────────────────────────────────────────

  /**
   * @param {number} bpm
   * @param {number} [atNtpSeconds=0] — honoured by a Link backing
   */
  setBpm(bpm, atNtpSeconds = 0) {
    this.#localBpm = bpm;
    if (this.#sabBigInt) {
      Atomics.store(this.#sabBigInt, SC_BPM_I64, doubleToBits(bpm));
    } else if (this.#workletPort) {
      this.#workletPort.postMessage({
        type: SuperClockMessageType.SET_SESSION_BPM,
        bpm, atNtpSeconds,
      });
    }
  }

  /**
   * @param {boolean} playing
   * @param {number} [atNtpSeconds=0]
   */
  setIsPlaying(playing, atNtpSeconds = 0) {
    this.#localIsPlaying = !!playing;
    this.#localIsPlayingAtNtp = atNtpSeconds;
    if (this.#sabBigInt) {
      Atomics.store(this.#sabBigInt, SC_IS_PLAYING_AT_NTP_I64, doubleToBits(atNtpSeconds));
      Atomics.store(this.#sabInt32,  SC_IS_PLAYING_I32, playing ? 1 : 0);
    } else if (this.#workletPort) {
      this.#workletPort.postMessage({
        type: SuperClockMessageType.SET_SESSION_IS_PLAYING,
        isPlaying: this.#localIsPlaying, atNtpSeconds,
      });
    }
  }

  /** @param {boolean} enabled */
  setLinkEnabled(enabled) {}

  /**
   * @param {number} beat
   * @param {number} atNtpSeconds
   * @param {number} quantum
   */
  requestBeatAtTime(beat, atNtpSeconds, quantum) {
    const bpm = this.getBpm();
    const newOrigin = atNtpSeconds - (beat * 60.0) / bpm;
    this.#localBeatOriginNtp = newOrigin;
    if (this.#sabBigInt) {
      Atomics.store(this.#sabBigInt, SC_BEAT_ORIGIN_NTP_I64, doubleToBits(newOrigin));
    } else if (this.#workletPort) {
      this.#workletPort.postMessage({
        type: SuperClockMessageType.SET_SESSION_BEAT_ORIGIN_NTP,
        beatOriginNtp: newOrigin,
      });
    }
  }

  forceBeatAtTime(beat, atNtpSeconds, quantum) {
    this.requestBeatAtTime(beat, atNtpSeconds, quantum);
  }

  // ── Session getters ────────────────────────────────────────────────────

  getBpm() {
    if (this.#sabBigInt) return bitsToDouble(Atomics.load(this.#sabBigInt, SC_BPM_I64));
    return this.#localBpm;
  }

  isPlaying() {
    if (this.#sabInt32) return Atomics.load(this.#sabInt32, SC_IS_PLAYING_I32) !== 0;
    return this.#localIsPlaying;
  }

  getBeatOriginNtp() {
    if (this.#sabBigInt) return bitsToDouble(Atomics.load(this.#sabBigInt, SC_BEAT_ORIGIN_NTP_I64));
    return this.#localBeatOriginNtp;
  }

  getIsPlayingAtNtp() {
    if (this.#sabBigInt) return bitsToDouble(Atomics.load(this.#sabBigInt, SC_IS_PLAYING_AT_NTP_I64));
    return this.#localIsPlayingAtNtp;
  }

  isLinkEnabled() { return false; }
  numPeers()      { return 0; }

  /**
   * Current NTP time as seen by the audio thread — `AudioContext.currentTime`
   * + the engine's NTP-start anchor + drift + global offset.
   *
   * Use this to schedule events relative to "now in audio time":
   *
   *   sonic.sendOSC(osc.encodeBundle(sonic.superClock.now() + 0.05, packets));
   *
   * That bundle fires 50 ms of *audio time* from now. The audio thread is
   * reading the same clock, so the two sides always agree — no skew if
   * AudioContext is throttled, no need to know the NTP-start formula.
   */
  now() {
    if (!this.#audioContext) return 0;
    return this.nowAt(this.#audioContext.currentTime);
  }

  /**
   * Compute audio-thread NTP for a specific `AudioContext.currentTime`.
   * `now()` calls this with the live `currentTime`; advanced callers can
   * pass a specific value (e.g. from `audioContext.getOutputTimestamp()`
   * for tight sample-aligned scheduling).
   */
  nowAt(audioCurrentTime) {
    // getDriftOffset / getClockOffset both return milliseconds; divide
    // by 1000 to get seconds. (C++ reads raw microseconds from the SAB
    // and divides by 1e6, getting the same seconds value with µs
    // precision — JS-side precision is ms because NTPTiming rounds.)
    return audioCurrentTime
         + this.getNTPStartTime()
         + this.getDriftOffset() / 1000
         + this.getClockOffset() / 1000;
  }

  /**
   * Current NTP time as seen by the system wall clock — `performance.now()`
   * + `performance.timeOrigin`, converted to NTP. Independent of the audio
   * clock; useful when matching against external wall-clock events (e.g.
   * network MIDI, NTP sources). For scheduling engine events, prefer
   * {@link now} so the audio thread and the scheduler agree.
   */
  wallNow() {
    return (performance.timeOrigin + performance.now()) / 1000 + 2208988800;
  }

  // ── Beat math ──────────────────────────────────────────────────────────
  // Reads bpm and beat_origin individually — no multi-field coherence
  // guarantee. Today's only callers are app-thread (single-writer JS).

  beatAtTime(ntpSeconds, quantum) {
    return (ntpSeconds - this.getBeatOriginNtp()) * this.getBpm() / 60.0;
  }

  phaseAtTime(ntpSeconds, quantum) {
    const beat = this.beatAtTime(ntpSeconds, quantum);
    let phase = beat % quantum;
    if (phase < 0) phase += quantum;
    return phase;
  }

  timeAtBeat(beat, quantum) {
    return this.getBeatOriginNtp() + beat * 60.0 / this.getBpm();
  }
}
